include/swift/Remote/MetadataReader.h

```diff
@@ -2905,25 +2909,50 @@ class MetadataReader {
     return builtSubsts;
   }
 
-  BuiltType readNominalTypeFromMetadata(MetadataRef origMetadata,
-                                        bool skipArtificialSubclasses = false) {
-    auto metadata = origMetadata;
-    auto descriptorAddress =
-      readAddressOfNominalTypeDescriptor(metadata,
-                                         skipArtificialSubclasses);
-    if (!descriptorAddress)
+  /// "Generic" version of readNominalTypeFromDescriptor when BuilderType !=
+  /// TypeRefBuilder.
+  template <
+      typename T = BuilderType,
+      std::enable_if_t<!std::is_same<T, reflection::TypeRefBuilder>::value,
+                       bool> = true>
+  BuiltType readNominalTypeFromDescriptor(MetadataRef metadata,
+                                          ContextDescriptorRef descriptor) {
+    if (!descriptor)
       return BuiltType();
 
-    // If we've skipped an artificial subclasses, check the cache at
-    // the superclass.  (This also protects against recursion.)
-    if (skipArtificialSubclasses && metadata != origMetadata) {
-      auto it = TypeCache.find(getAddress(metadata));
-      if (it != TypeCache.end())
-        return it->second;
+    // From that, attempt to resolve a nominal type.
+    BuiltTypeDecl typeDecl = buildNominalTypeDecl(descriptor);
+    if (!typeDecl)
+      return BuiltType();
+
+    // Build the nominal type.
+    BuiltType nominal;
+    if (descriptor->isGeneric()) {
+      // Resolve the generic arguments.
+      auto builtGenerics = getGenericSubst(metadata, descriptor);
+      if (builtGenerics.empty())
+        return BuiltType();
+      nominal = Builder.createBoundGenericType(typeDecl, builtGenerics);
+    } else {
+      nominal = Builder.createNominalType(typeDecl);
     }
 
-    // Read the nominal type descriptor.
-    ContextDescriptorRef descriptor = readContextDescriptor(descriptorAddress);
+    return nominal;
+  }
+
+  /// Specialized version of readNominalTypeFromDescriptor where BuilderType ==
+  /// TypeRefBuilder. There are two main differences in this specialization:
+  /// 1 - We attempt to rebuild the parent typeref and pass that along
+  /// when creating our own typeref.
+  /// 2 - We remove the generic parameters that belong to our typeref's parents
+  /// from the flat array of generic parameters we read from the context
+  /// descriptor, as typerefs store generic parameters per level instead of as a
+  /// single flat array.
+  template <typename T = BuilderType,
+            std::enable_if_t<std::is_same<T, reflection::TypeRefBuilder>::value,
+                             bool> = true>
```

Comment:
Instead of hardcoding `TypeRefBuilder` here, it might be cleaner to look for a new trait on `BuilderType`, so we can do `enable_if_t<BuilderType::needsToPrecomputeParentGenericContextShapes>` or something like that.